src/java/org/apache/cassandra/index/sai/disk/v1/MemtableIndexWriter.java

```diff
@@ -0,0 +1,177 @@
+
+    public MemtableIndexWriter(MemtableIndex memtable,
+                               IndexDescriptor indexDescriptor,
+                               IndexContext indexContext,
+                               RowMapping rowMapping)
+    {
+        assert rowMapping != null && rowMapping != RowMapping.DUMMY : "Row mapping must exist during FLUSH.";
+
+        this.indexDescriptor = indexDescriptor;
+        this.indexContext = indexContext;
+        this.memtable = memtable;
+        this.rowMapping = rowMapping;
+    }
+
+    @Override
+    public void addRow(PrimaryKey key, Row row, long sstableRowId)
+    {
+        // Memtable indexes are flushed directly to disk with the aid of a mapping between primary
+        // keys and row IDs in the flushing SSTable. This writer, therefore, does nothing in
+        // response to the flushing of individual rows.
+    }
+
+    @Override
+    public void abort(Throwable cause)
+    {
+        logger.warn(indexContext.logMessage("Aborting index memtable flush for {}..."), indexDescriptor.descriptor, cause);
+        indexDescriptor.deleteColumnIndex(indexContext);
+    }
+
+    @Override
+    public void complete(Stopwatch stopwatch) throws IOException
+    {
+        long start = stopwatch.elapsed(TimeUnit.MILLISECONDS);
+
+        try
+        {
+            if (!rowMapping.hasRows() || (memtable == null) || memtable.isEmpty())
+            {
+                logger.debug(indexContext.logMessage("No indexed rows to flush from SSTable {}."), indexDescriptor.descriptor);
+                // Write a completion marker even though we haven't written anything to the index,
+                // so we won't try to build the index again for the SSTable
+                indexDescriptor.createComponentOnDisk(IndexComponent.COLUMN_COMPLETION_MARKER, indexContext);
+                return;
+            }
+
+            final DecoratedKey minKey = rowMapping.minKey.partitionKey();
+            final DecoratedKey maxKey = rowMapping.maxKey.partitionKey();
+
+            final Iterator<Pair<ByteComparable, LongArrayList>> iterator = rowMapping.merge(memtable);
+
+            try (MemtableTermsIterator terms = new MemtableTermsIterator(memtable.getMinTerm(), memtable.getMaxTerm(), iterator))
+            {
+                long cellCount = flush(minKey, maxKey, indexContext.getValidator(), terms);
+
+                indexDescriptor.createComponentOnDisk(IndexComponent.COLUMN_COMPLETION_MARKER, indexContext);
+
+                indexContext.getIndexMetrics().memtableIndexFlushCount.inc();
+
+                long elapsed = stopwatch.elapsed(TimeUnit.MILLISECONDS);
+
+                logger.debug(indexContext.logMessage("Completed flushing {} memtable index cells to SSTable {}. Duration: {} ms. Total elapsed: {} ms"),
+                             cellCount,
+                             indexDescriptor.descriptor,
+                             elapsed - start,
+                             elapsed);
```

Comment:
style/nit: I actually like putting these all on one line, because they sort of line up w/ the `{}` above :)

Reply:
I'll probably get shot down by someone else for not following formatting standards mind :)